doc/xl2tpd-control.8
.TH xl2tpd-control 8 "Jan 2022"

.SH NAME
xl2tpd\-control \- xl2tpd control utility.

.SH DESCRIPTION
xl2tpd is an implementation of the Layer 2 Tunneling Protocol (RFC 2661).
L2TP allows to tunnel PPP over UDP. Some ISPs use L2TP to tunnel user sessions
from dial-in servers (modem banks, ADSL DSLAMs) to back-end PPP servers.
Another important application is Virtual Private Networks where the IPsec
protocol is used to secure the L2TP connection (L2TP/IPsec, RFC 3193).

xl2tpd works by opening a pseudo-tty for communicating with pppd.
It runs completely in userspace but supports kernel mode L2TP.

xl2tpd supports IPsec SA Reference tracking to enable overlapping internak
NAT'ed IP's by different clients (eg all clients connecting from their
linksys internal IP 192.168.1.101) as well as multiple clients behind
the same NAT router.

.SH SYNOPSIS
.HP
\fBxl2tpd-control\fR [\fI-c\fR PATH] \fI<COMMAND>\fR \fI<TUNNEL_NAME>\fR [\fIOPTIONS\fR]


.SH OPTIONS
.TP
.B -c
This option specifies xl2tpd control file.

.TP
.B -d
This option enables debugging mode.

.SH COMMANDS
.TP
.B add-lac
Adds new or modify existing LAC (L2TP Access Concentrator) configuration.
Configuration should be specified as a command options in <key>=<value>
pairs format. See available options in xl2tpd.conf(5).

.TP
.B connect-lac
Establish new connection to LAC.
Username and secret for tunnel can be passed as a command options.

.TP
.B disconnect-lac
Disconnects tunnel.

.TP
.B remove-lac
Removes existing LAC configuration.
xl2tpd disconnects the tunnel before removing.

.TP
.B add-lns
Adds new or modify existing LNS (L2TP Network Server) configuration.

.TP
.B remove-lns
Removes existing LNS configuration.

.TP
.B status-lns
Check the status of LNS.

.TP
.B available
Check availability.

.SH BUGS
Please use the github project page
https://github.com/xelerance/xl2tpd
to send bugreports, issues and any other feedback.

.SH SEE ALSO
xl2tpd.conf(5),
xl2tpd(8),
pppd(8)

.SH COPYLEFT
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (see the file LICENSE); if not, see
https://www.gnu.org/licenses/, or contact Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

.SH CONTRIBUTORS
Alexander Dorokhov <alex.dorokhov@gmail.com>
.br
Alexander Naumov <alexander_naumov@opensuse.org>

.SH AUTHORS
Forked from l2tpd by Xelerance: https://github.com/xelerance/xl2tpd

Michael Richardson <mcr@xelerance.com>
.br
Paul Wouters <paul@xelerance.com>
.br
Samir Hussain <shussain@xelerance.com>

Previous development was hosted at sourceforge
(http://www.sourceforge.net/projects/l2tpd) by:
.P
Scott Balmos <sbalmos@iglou.com>
.br
David Stipp <dstipp@one.net>
.br
Jeff McAdams <jeffm@iglou.com>

Based off of l2tpd version 0.61.
Many thanks to Jacco de Leeuw <jacco2@dds.nl> for maintaining l2tpd.

.br
Copyright (C)1998 Adtran, Inc.
.br
Mark Spencer <markster@marko.net>


CHANGES
v1.3.18dev1 (Jan 20, 2022)

v1.3.17 (Jan 20, 2022)
* debian: Add systemd .service [Will Thompson]
* Update TODO file to specify branch work that needs to be completed [Samir Hussain]
* Correct a typo in the man page xl2tpd.conf(5) [Samy Mahmoudi]
* xl2tpd-control: fix typo in usage [Yousong Zhou]
* Updating debian changelog description [Samir Hussain]
* control: Fix branch of Vcs-Git [Samuel Thibault]
* upload [Samuel Thibault]
* Drop unused lintian warning that produces a lintian error [Samuel Thibault]
* Update xl2tpd.doc to use README.md [Samir Hussain]
* Drop patch that cannot be applied any more [Samuel Thibault]
* Bump debhelper from 10 to 12 [Samuel Thibault]
* control: Bump Standards-Version to 4.5.0 [Samuel Thibault]
* Really move changes to patch series [Samuel Thibault]
* Move c0cce208e4a5 to patch series [Samuel Thibault]
* upload [Samuel Thibault]
* Close bug [Samuel Thibault]
* Use https URL. [Samuel Thibault]

v1.3.16 (October 23, 2020)
* Re-add braces for if-else that have only statement [Samir Hussain]
* xl2tpd-control refactoring [Alexander Naumov]
* fix travis 'script' syntax [Alexander Naumov]
* adding xl2tpd-control tests to travis [Alexander Naumov]
* Re-adding text giving more inofrmation about using ipsec with xl2tpd [Samir Hussain]
* Update README: typo, links to RFC, link to travis [Alexander Naumov]
* Travis will test different compiler on linux distro [Samir Hussain]
* Update travis for proper matrix [Samir Hussain]
* Add Focal to travis testing [Samir Hussain]
* yet another man-page update [Alexander Naumov]
* update man-pages, fix typo [Alexander Naumov]
* set_flow: result of operation is garbage or undefined [Alexander Naumov]
* Update README [Alexander Naumov]
* adding xl2tpd.init.patch [Alexander Naumov]
* adding Makefile.patch [Alexander Naumov]
* moving changes to separate file [Alexander Naumov]
* sync/update spec file with official SUSE version [Alexander Naumov]
* Set IP_PKTINFO even if setting of IPPROTO_IP fails (as it was in 1.3.9) [shadyhh]
* Update .gitignore vim swap file [Samir Hussain]
* Travis will test supported Ubuntu LTS distros [Samir Hussain]
* Updating COMPATABILITY_ISSUES with info on Miktrotik servers [Samir Hussain]
* Add work around for Android 10 maxium retries in COMPATIBILITY_ISSUES [Samir Hussain]
* Add compatability issues with Ciso ASA [Samir Hussain]

v1.3.15 (October 13, 2019)
* Fix spacing of CONTRIBUTION.md [Samir Hussain]
* Add CONTRIBUTION.md [Samir Hussain]
* Specify missing log arguments [Patch by github user: 川島和津実]
* Use matrix for .travis.yaml to test for multiple Linux distro [Samir Hussain]
* Fixing .travis.yaml spacing warning [Samir Hussain]
* Sockopt bug fix for multiple IP's [JDTX]
* Add Clang as compiler test for travis [Samir Hussain]
* Add info on building and installing xl2tpd [Samir Hussain]

v1.3.14 (April 17, 2019)
* osport.h: replace SUSv3-specific functions by POSIX variants [Fabrice Fontaine]
* avp: Error Code field in Result Code AVP is optional [Pau Espin Pedrol]
* network_thread: Early continue in loop to remove huge indented block [Pau Espin Pedrol]
*  network_thread: Simplify while loop using for loop [Pau Espin Pedrol]
* network: connect_pppol2tp: early return to avoid huge indentation block [Pau Espin Pedrol]
* xl2tpd: start_pppd: Fix truncation of last character [Pau Espin Pedrol]
* handle_packet: Remove unneded else clause when handling payload [Pau Espin Pedrol]
* control: Split control message handling into its own function [Pau Espin Pedrol]
* handle_packet: Rearrange code flow to simplify it [Pau Espin Pedrol]
* avp: Early failure if no handler to remove indent block [Pau Espin Pedrol]
* xl2tpd: Mark internal symbols as static [Pau Espin Pedrol]
* Fix indentation and whitespace in code block [Pau Espin Pedrol]
* xl2tpd: Remove unused variable [Pau Espin Pedrol]
* network: Add missing close(kernel_fd) on init network failure [Pau Espin Pedrol]
* network: Add missing close(server_fd) on init network failure [Pau Espin Pedrol]
* Add 'cap backoff' option, limiting exponential backoff retries will
  be delayed by exponentially longer time, unless that time is capped
  by configuration. [Bart Trojanowski]
* Add program to show status icon in system tray. [Github user: username34]
* Add info on building and installing xl2tpd [Samir Hussain]
* Update formatting of README.md [Samir Hussain]
* Rename README.xl2tpd to README.md [Samir Hussain]
* Update Debian changelog [Samir Hussain]

v1.3.13 (December 3, 2018)
* Specify email address for reporting security vulnerabilities [Samir Hussain]
* Fix compile warning with USE_KERNEL in xl2tpd.c [Samir Hussain]
* Applying patch that reduces compile warnings and fixes warnings from gcc and clang. [Gareth Ansell]
* Fix compiler warnings in network.c [Gareth Ansell]
* Add a make command for packaging's prep work [Samir Hussain]
* Add Makefile directive for getting version [Samir Hussain]
* Add a preproc for Watchguard firewall (Github issue #136) [daniel1111]
* Convert from ISO-8859 to UTF-8 [Simon Deziel]
* Update README to provide latest info on xl2tpd + Linux kernel 4.15+ [Samir Hussain]
* Use dh_auto_build in order to allow cross compiles [Helmut Grohne]

v1.3.12 (May 18, 2018)
* TOS value to copy to the tunnel header (Yurkovskyy)
* Fix for ENODEV (No such device) error with Linux kernel 4.15 (Douglas Kosovic)
* Update xl2tpd.init (bogdik)
* fix version number and upload (Samuel Thibault)
* import wheezy changes  (Samuel Thibault)

v1.3.11 (March 7, 2018)
* Build packages for Xenial by default (Simon Deziel)
* Bump d/compat to 9 (Simon Deziel)
* Drop d/repack.sh script and refresh d/watch (Simon Deziel)
* Refresh d/control by partly sync'ing from Debian (Simon Deziel)
* Use HTTPS URL in d/copyright (Simon Deziel)

v1.3.10.1 (November 16, 2017)
* Have max retries as a configuration [Samir Hussain]
* Add more into to "select timeout" debug message [Samir Hussain]

v1.3.10 (August 2, 2017)
* Update STRLEN in file.h to 100 (from 80) [Samir Hussain]

xl2tpd-control.c
/*
 * xl2tpd-control - the xl2tpd control utility
 *
 * Copyright (C) 2011 Alexander Dorokhov <alex.dorokhov@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (see the file LICENSE); if not, see
 * https://www.gnu.org/licenses/, or contact Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 ****************************************************************
 */
 
#define _GNU_SOURCE

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <errno.h>

#include "l2tp.h"

/* Paul: Alex: can we change this to use stdout, and let applications using
 * xl2tpd-control capture the output, instead of creating tmp files?
 */
/* result filename format including absolute path and formatting %i for pid */
#define RESULT_FILENAME_FORMAT "/var/run/xl2tpd/xl2tpd-control-%i.out"

#define ERROR_LEVEL 1
#define DEBUG_LEVEL 2

#define TUNNEL_REQUIRED 1
#define TUNNEL_NOT_REQUIRED 0
#define TIMEOUT 1000000  //timeout is 1s

char result_filename[128];
int result_fd = -1;

int log_level = ERROR_LEVEL;

void print_error (int level, const char *fmt, ...);
int read_result(int result_fd, char* buf, ssize_t size);

/* Definition of a command */
struct command_t
{
    char *name;
    int (*handler) (FILE*, char* tunnel, int optc, char *optv[]);
    int requires_tunnel;
};

int command_add_lac       (FILE*, char* tunnel, int optc, char *optv[]);
int command_connect_lac   (FILE*, char* tunnel, int optc, char *optv[]);
int command_disconnect_lac(FILE*, char* tunnel, int optc, char *optv[]);
int command_status_lac    (FILE*, char* tunnel, int optc, char *optv[]);
int command_remove_lac    (FILE*, char* tunnel, int optc, char *optv[]);
int command_available     (FILE*, char* tunnel, int optc, char *optv[]);
int command_add_lns       (FILE*, char* tunnel, int optc, char *optv[]);
int command_status_lns    (FILE*, char* tunnel, int optc, char *optv[]);
int command_remove_lns    (FILE*, char* tunnel, int optc, char *optv[]);

struct command_t commands[] = {
    {"add-lac",       &command_add_lac,       TUNNEL_REQUIRED},
    {"connect-lac",   &command_connect_lac,   TUNNEL_REQUIRED},
    {"disconnect-lac",&command_disconnect_lac,TUNNEL_REQUIRED},
    {"status-lac",    &command_status_lac,    TUNNEL_REQUIRED},
    {"remove-lac",    &command_remove_lac,    TUNNEL_REQUIRED},
    {"available",     &command_available,     TUNNEL_NOT_REQUIRED},
    {"add-lns",       &command_add_lns,       TUNNEL_REQUIRED},
    {"status-lns",    &command_status_lns,    TUNNEL_REQUIRED},
    {"remove-lns",    &command_remove_lns,    TUNNEL_REQUIRED}
};

void usage()
{
    printf ("Usage: xl2tpd-control [-c <PATH>] <COMMAND> <TUNNEL_NAME> [<OPTIONS>]\n\n"
            "       -c  set xl2tpd control file\n"
            "       -d  enable debugging mode\n"
            "--version  show version\n"
            "   --help  show this help message\n\n"
            "List of supported commands:\n"
            "add-lac, status-lac, remove-lac, connect-lac, disconnect-lac\n"
            "add-lns, status-lns, remove-lns, avaliable\n\n"
            "See xl2tpd-control(8) man page for more details.\n");
}

void cleanup(void)
{
    /* cleaning up */
    unlink (result_filename);
    if (result_fd >= 0)
        close (result_fd);
}

int main (int argc, char *argv[])
{
    char* control_filename = NULL;
    char* tunnel_name = NULL;
    struct command_t* command = NULL;    
    int i; /* argv iterator */

    if (argv[1] && !strncmp (argv[1], "--help", 6)) {
        usage();
        return 0;
    }

    if (argv[1] && !strncmp (argv[1], "--version", 9)) {
        printf ("Version: %s\n", SERVER_VERSION);
        return 0;
    }

    /* parse global options */
    for (i = 1; i < argc; i++) {
        if (!strncmp (argv[i], "-c", 2)) {
            control_filename = argv[++i];
        } else if (!strncmp (argv[i], "-d", 2)) {
            log_level = DEBUG_LEVEL;
        } else {
            break;
        }
    }

    if (i >= argc) {
        print_error (ERROR_LEVEL, "error: command not specified\n");
        usage();
        return -1;
    }

    if (!control_filename) {
        control_filename = strdup (CONTROL_PIPE);
    }

    /* parse command name */
    for (command = commands; command->name; command++)
    {
        if (!strcasecmp (argv[i], command->name))
        {
            i++;
            break;
        }
    }
    
    if (!command->name) {
        print_error (ERROR_LEVEL, "error: no such command\n");
        free(control_filename);
        usage();
        return -1;
    }
    
    /* get tunnel name */
    if(command->requires_tunnel){
        if (i >= argc) {
            print_error (ERROR_LEVEL, "error: tunnel name not specified\n");
            usage();
            free(control_filename);
            return -1;
        }
        tunnel_name = argv[i++];    
        /* check tunnel name for whitespaces */
        if (strstr (tunnel_name, " ")) {
            print_error (ERROR_LEVEL, "error: tunnel name shouldn't include spaces\n");
            usage();
            free(control_filename);
            return -1;
        }
    }
    
    char buf[CONTROL_PIPE_MESSAGE_SIZE] = "";
    FILE* mesf = fmemopen (buf, CONTROL_PIPE_MESSAGE_SIZE, "w");

    /* create result pipe for reading */
    snprintf (result_filename, 128, RESULT_FILENAME_FORMAT, getpid());
    unlink (result_filename);
    mkfifo (result_filename, 0600);
    atexit(cleanup);

    result_fd = open (result_filename, O_RDONLY | O_NONBLOCK, 0600);
    if (result_fd < 0)
    {
        print_error (ERROR_LEVEL,
            "error: unable to open %s for reading.\n", result_filename);
        return -2;
    }
   
    /* turn off O_NONBLOCK */
    if (fcntl (result_fd, F_SETFL, O_RDONLY) == -1) {
        print_error (ERROR_LEVEL,
            "Can not turn off nonblocking mode for result_fd: %s\n",
            strerror(errno));
        return -2;
    }
    
    /* pass result filename to command */
    fprintf (mesf, "@%s ", result_filename);
    if (ferror (mesf)) {
        print_error (ERROR_LEVEL, "internal error: message buffer to short");
        return -2;
    }
    
    /* format command with remaining arguments */
    int command_res = command->handler (
        mesf, tunnel_name, argc - i, argv + i
    );

    if (command_res < 0) {
        print_error (ERROR_LEVEL, "error: command parse error\n");
        return -1;
    }
    
    fflush (mesf);
    
    if (ferror (mesf)) {
        print_error (ERROR_LEVEL,
            "error: message too long (max = %i ch.)\n",
            CONTROL_PIPE_MESSAGE_SIZE - 1);
        return -1;
    }
    
    print_error (DEBUG_LEVEL, "command to be passed:\n%s\n", buf);

    /* try to open control file for writing */
    int control_fd = open (control_filename, O_WRONLY | O_NONBLOCK, 0600);
    if (control_fd < 0) {
        int errorno = errno;
        switch (errorno)
        {
        case EACCES:
            print_error (ERROR_LEVEL,
                "Unable to open %s for writing."
                " Is xl2tpd running and you have appropriate permissions?\n",
                control_filename);
            break;
        default:
            print_error (ERROR_LEVEL,
                "Unable to open %s for writing: %s\n",
                control_filename, strerror (errorno));
        }
        return -1;
    }

    /* turn off O_NONBLOCK */
    if (fcntl (control_fd, F_SETFL, O_WRONLY) == -1) {
        print_error (ERROR_LEVEL,
            "Can not turn off nonblocking mode for control_fd: %s\n",
            strerror(errno));
        return -2;
    }
    
    /* pass command to control pipe */
    if (write (control_fd, buf, ftell (mesf)) < 0) {
      int errorno = errno;
      print_error (ERROR_LEVEL,
                "Unable to write to %s: %s\n",
                control_filename, strerror (errorno));
      close (control_fd);
      return -1;
    }
    close (control_fd);
    
    /* read result from pipe */
    char rbuf[CONTROL_PIPE_MESSAGE_SIZE] = "";
    int command_result_code = read_result (
        result_fd, rbuf, CONTROL_PIPE_MESSAGE_SIZE
    );
    /* rbuf contains a newline, make it double to form a boundary. */
    print_error (DEBUG_LEVEL, "command response: \n%s\n", rbuf);
    
    return command_result_code;
}

void print_error (int level, const char *fmt, ...)
{
    if (level > log_level) return;
    va_list args;
    va_start (args, fmt);
    fprintf (stderr, "xl2tpd-control: ");
    vfprintf (stderr, fmt, args);
    va_end (args);
}


int read_result(int result_fd, char* buf, ssize_t size)
{
    /* read result from result_fd */
    /*FIXME: there is a chance to hang up reading.
             Should I create watching thread with timeout?
     */
    ssize_t readed = 0;
    ssize_t len;
    int write_pipe = 0;
    struct timeval tvs;
    struct timeval tve;
    unsigned long diff;
    gettimeofday(&tvs, NULL);

    do
    {
        len = read (result_fd, buf + readed, size - readed);
        if (len < 0)
        {
            if (errno == EINTR)
                continue;
            print_error (ERROR_LEVEL,
                "error: can't read command result: %s\n", strerror (errno));
            break;
        }
        else if (len == 0) {
            if(!write_pipe) {
                 gettimeofday(&tve, NULL);
                 diff = (tve.tv_sec - tvs.tv_sec) * 1000000 + (tve.tv_usec - tvs.tv_usec);
                 if (diff >= TIMEOUT) {
                     print_error (DEBUG_LEVEL, "error: read timout\n");
                     break;
                 } else {
                     usleep(10);
                     continue;
                 }
            }
            break;
        }
        else {
            write_pipe = 1;
            readed += len;
            if ((size - readed) <= 0)
                break;
       }
    } while (1);

    buf[readed] = '\0';

    /* scan result code */
    int command_result_code = -3;
    sscanf (buf, "%i", &command_result_code);

    return command_result_code;
}

int command_add
(FILE* mesf, char* tunnel, int optc, char *optv[], int reqopt)
{
    if (optc <= 0) {
        print_error (ERROR_LEVEL, "error: tunnel configuration expected\n");
        return -1;
    }

    fprintf (mesf, "%c %s ", reqopt, tunnel);
    int i;
    int wait_key = 1;
    for (i = 0; i < optc; i++)
    {
        fprintf (mesf, "%s", optv[i]);
        if (wait_key)
        {
            /* try to find '=' */
            char* eqv = strstr (optv[i], "=");
            if (eqv) {
                /* check is it not last symbol */
                if (eqv != (optv[i] + strlen(optv[i]) - 1)) {
                    fprintf (mesf, ";"); /* end up option */
                } else {
                    wait_key = 0; /* now we waiting for value */
                }
            } else { /* two-word key */
                fprintf (mesf, " "); /* restore space */
            }
        } else {
            fprintf (mesf, ";"); /* end up option */
            wait_key = 1; /* now we again waiting for key */
        }
    }
    return 0;
}

int command_add_lac
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    return command_add(mesf, tunnel, optc, optv, CONTROL_PIPE_REQ_LAC_ADD_MODIFY);
}

int command_add_lns
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    return command_add(mesf, tunnel, optc, optv, CONTROL_PIPE_REQ_LNS_ADD_MODIFY);
}


int command_connect_lac
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_LAC_CONNECT, tunnel);
    /* try to read authname and password from opts */
    if (optc > 0) {
        if (optc == 1)
            fprintf (mesf, " %s", optv[0]);
        else // optc >= 2
            fprintf (mesf, " %s %s", optv[0], optv[1]);
    }
    return 0;
}

int command_disconnect_lac
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    UNUSED(optc);
    UNUSED(optv);
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_LAC_DISCONNECT, tunnel);
    return 0;
}

int command_remove_lac
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    UNUSED(optc);
    UNUSED(optv);
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_LAC_REMOVE, tunnel);
    return 0;
}

int command_status_lns
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    UNUSED(optc);
    UNUSED(optv);
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_LNS_STATUS, tunnel);
    return 0;
}

int command_status_lac
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    UNUSED(optc);
    UNUSED(optv);
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_LAC_STATUS, tunnel);
    return 0;
}

int command_available
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    UNUSED(optc);
    UNUSED(optv);
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_AVAILABLE, tunnel);
    return 0;
}

int command_remove_lns
(FILE* mesf, char* tunnel, int optc, char *optv[])
{
    UNUSED(optc);
    UNUSED(optv);
    fprintf (mesf, "%c %s", CONTROL_PIPE_REQ_LNS_REMOVE, tunnel);
    return 0;
}

